File: 00-code(源代码)/src/com/hi/dhl/algorithms/_155/java/MinStack.java
package com.hi.dhl.algorithms._155.java;

import java.util.ArrayDeque;

/**
 * <pre>
 *     author: dhl
 *     date  : 2020/8/9
 *     desc  :
 * </pre>
 */
class MinStack {
    ArrayDeque<Integer> stack;
    ArrayDeque<Integer> minStack;

    /**
     * initialize your data structure here.
     */
    public MinStack() {
        stack = new ArrayDeque<Integer>();
        minStack = new ArrayDeque<Integer>();
    }

    public void push(int x) {
        stack.push(x);
        if (minStack.peek() == null) {
            minStack.push(x);
        } else {
            minStack.push(Math.min(x, minStack.peek()));
        }
    }

    public void pop() {
        stack.poll();
        minStack.poll();
    }

    public int top() {
        return stack.peek();
    }

    public int getMin() {
        return minStack.peek();
    }
}

File: 00-code(源代码)/src/com/hi/dhl/algorithms/leetcode/_155/kotlin/MinStack.kt
package com.hi.dhl.algorithms.leetcode._155.kotlin

import java.util.*

/**
 * <pre>
 *     author: dhl
 *     date  : 2020/8/9
 *     desc  :
 * </pre>
 */

class MinStack() {

    /** initialize your data structure here. */
    val stack = ArrayDeque<Int>()
    val minStack = ArrayDeque<Int>()

    fun push(x: Int) {
        stack.push(x)
        if (minStack.peek() == null) {
            minStack.push(x)
        } else {
            minStack.push(Math.min(x, minStack.peek()))
        }
    }

    fun pop() {
        stack.poll()
        minStack.poll()
    }

    fun top(): Int {
        return stack.peek()
    }

    fun getMin(): Int {
        return minStack.peek()
    }

}